lc0/src/neural/network_opencl.cc

```diff
@@ -0,0 +1,447 @@
+/*
+ This file is part of Leela Chess Zero.
+ Copyright (C) 2018 The LCZero Authors
+ 
+ Leela Chess is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+ 
+ Leela Chess is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+ 
+ You should have received a copy of the GNU General Public License
+ along with Leela Chess.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "neural/factory.h"
+#include "neural/network.h"
+
+#include <condition_variable>
+#include <cassert>
+#include <iostream>
+#include <cmath>
+#include <algorithm>
+#include <thread>
+#include "utils/exception.h"
+
+#include "CL/OpenCLUtils.h"
+
+#include "CL/OpenCLParams.h"
+#include "CL/OpenCL.h"
+
+#ifdef __APPLE__
+#include <Accelerate/Accelerate.h>
+#endif
+#ifdef USE_MKL
+#include <mkl.h>
+#endif
+#ifdef USE_OPENBLAS
+#include <cblas.h>
+#endif
+
+
+namespace lczero {
+  
+  namespace {
+    
+    
+    static constexpr int NUM_VALUE_INPUT_PLANES = 32;
+    static constexpr int NUM_POLICY_INPUT_PLANES = 32;
+    
+    static constexpr int NUM_OUTPUT_POLICY = 1858;
+    static constexpr int NUM_VALUE_CHANNELS = 128;
+    
+    
+    class OpenCLNetwork;
+    
+    // Copy the vectors we need after weights is deallocated
+    struct OpenCLWeights {
+      
+      const std::vector<float> ip2_val_w;
+      const std::vector<float> ip2_val_b;
+      
+      OpenCLWeights(const Weights& weights):
+      ip2_val_w(weights.ip2_val_w),
+      ip2_val_b(weights.ip2_val_b)
+      {
+        
+      }
+
+    };
+    
+    
+    
+    class OpenCLComputation : public NetworkComputation {
+      
+    public:
+      
+      OpenCLComputation(const OpenCL_Network& opencl_net, const OpenCLWeights& weights):
```

Comment:
...shouldn't this be OpenCLNetwork? where's the random underscore from?

Reply:
This is the original class "OpenCL_Network" on OpenCL side that does the actual computation.
I did not change the original name ("OpenCL_Network"), as it is not really a misnaming.